haystack/preview/document_stores/memory/document_store.py

```diff
@@ -259,3 +270,108 @@ def bm25_retrieval(
+        :param top_k: The number of top documents to retrieve. Default is 10.
+        :param scale_score: Whether to scale the scores of the retrieved Documents. Default is True.
+        :param return_embedding: Whether to return the embedding of the retrieved Documents. Default is False.
+        :return: A list of the top 'k' documents most relevant to the query.
+        """
+        if (
+            not isinstance(query_embedding, list)
+            or len(query_embedding) == 0
+            or not isinstance(query_embedding[0], float)
+        ):
+            raise ValueError("query_embedding should be a non-empty list of floats.")
+
+        filters = filters or {}
+        all_documents = self.filter_documents(filters=filters)
+
+        documents_with_embeddings = [doc for doc in all_documents if doc.embedding is not None]
+        if len(documents_with_embeddings) == 0:
+            logger.info(
+                "No Documents found with embeddings. Returning empty list. "
+                "To generate embeddings, use a DocumentEmbedder."
+            )
+            return []
+        elif len(documents_with_embeddings) < len(all_documents):
+            logger.info(
+                "Skipping some Documents that don't have an embedding. "
+                "To generate embeddings, use a DocumentEmbedder."
+            )
+
+        # if statement in the list comprehension is needed to avoid a mypy error
+        embedding_sizes = [len(doc.embedding) for doc in documents_with_embeddings if doc.embedding is not None]
+        if any(size != embedding_sizes[0] for size in embedding_sizes):
+            raise ValueError(
+                "The embedding size of all Documents should be the same."
+                "Please make sure that the Documents have been embedded with the same model."
+            )
+
+        if len(query_embedding) != embedding_sizes[0]:
+            raise ValueError(
+                "The embedding size of the query should be the same as the embedding size of the Documents."
+                "Please make sure that the query has been embedded with the same model as the Documents."
+            )
+
+        scores = self._compute_embedding_similarity_scores(
+            query_embedding=query_embedding, documents=documents_with_embeddings, scale_score=scale_score
+        )
+
+        # create Documents with the similarity score for the top k results
+        top_documents = []
+        for doc, score in sorted(zip(documents_with_embeddings, scores), key=lambda x: x[1], reverse=True)[:top_k]:
+            doc_fields = doc.to_dict()
+            doc_fields["score"] = score
+            if return_embedding is False:
+                doc_fields["embedding"] = None
+            top_documents.append(Document(**doc_fields))
+
+        return top_documents
+
+    def _compute_embedding_similarity_scores(
+        self, query_embedding: List[float], documents: List[Document], scale_score: bool = True
+    ) -> List[float]:
```

Comment:
I would change a bit the naming here. `_compute_embedding_similarity_scores` to `_compute_query_embedding_similarity_scores` and `query_embedding` to `embeddings`.

This way we can rename `query_embedding_arr` in the body to `query_embeddings`, also `document_embeddings_arr` to `document_embeddings`.